src/main/java/com/marklogic/kafka/connect/sink/DefaultSinkRecordConverter.java

```diff
@@ -73,7 +105,29 @@ protected AbstractWriteHandle toContent(SinkRecord record) {
 			throw new NullPointerException("'record' must not be null, and must have a value.");
 		}
 		Object value = record.value();
+		Schema schema = record.valueSchema();
+		/* Determine the converter */
+		if (schema != null && value instanceof Struct) {
+			/* Avro or JSON with schema, ignore schema, handle only the value */
+			converter = "AVRO_OR_JSON_WITH_SCHEMA";
+			logger.info("Avro or JsonWithSchema document received. Converting to byte[].");
+			final String payload = new String(JSON_CONVERTER.fromConnectData(record.topic(), schema, value), StandardCharsets.UTF_8);
+			value = payload.getBytes();
+		}
+		
+        if (value instanceof Map) {
+        	converter = "JSON_WITHOUT_SCHEMA";
+        	logger.info("Json document received. Converting to byte[]");
+        	value = new String (JSON_CONVERTER.fromConnectData(record.topic(), null, value)).getBytes();
+        }
+        
+		if (value instanceof String) {
+			logger.info("String document received. Converting to byte[]");
+			converter = "STRING";
+        }
+		
 		if (value instanceof byte[]) {
+			logger.info("Byte[] received. Proceeding with MarkLogic operations.");
 			BytesHandle content = new BytesHandle((byte[]) value);
 			if (format != null) {
 				content.withFormat(format);
```

Comment:
Should this be two new implementations of the SinkRecordConverter class - one for JSON and one for AVRO? Then the correct converter would be specified in the config.

Reply:
SinkRecord has a schema part and a value part. For the messages having schema (JSON with schema, AVRO, ProtoBuf), the schema will be populated and value will be of type struct. The fromConnectData method of the kafka Connect converts the value to a String. For JSON (without Schema), value is of type Map. For everything else, the value is of type String or byte[]. 
For basically, the DefaultSinkRecordConverter is handling AVRO, ProtoBuf, JSON with schema, JSON and any other messages into format that is loadable to MarkLogic (ie byte[] with GenericDocumentManager). Schema portion of messages carrying schema are ignored. 
So, I think that the DefaultSinkRecordConverter has a common behavior of converting SinkRecord.value to byte[]. Other implementations in future can be to handle situations where SinkRecord.schema also need to be handled. That was my thinking and the user need not give which Converter to be used because for all messages the same thing is done - Convert SinkRecord.value to byte[]
If message type based implementations are provided, then we should give AVRO, ProtoBuf, JsonWithSchema, JsonWithoutSchema and String. Or a more general classification like MessagesWithSchema or MessagesWithoutSchema.
Let me know. 

ProtoBuf vs AVRO - They are different types of message formats and significantly differ in schema management. 

A protobuf and AVRO schemas for the same JSON is given below 

**JSON**
{"Snack":{"name":"cake","calories":260,"colour":"white", "flavor":"vanilla"}}
**ProtoBuf Schema**
`syntax = "proto3";
message Snack {
message obj {
string name = 1;
double calories = 2;
string colour = 3;
string flavor = 4;
}
obj Snack = 1;
}`

**AVRO Schema**
`{
  "name": "MyClass",
  "type": "record",
  "namespace": "com.acme.avro",
  "fields": [
    {
      "name": "Snack",
      "type": {
        "name": "Snack",
        "type": "record",
        "fields": [
          {
            "name": "name",
            "type": "string"
          },
          {
            "name": "calories",
            "type": "int"
          },
          {
            "name": "colour",
            "type": "string"
          },
          {
            "name": "flavor",
            "type": "string"
          }
        ]
      }
    }
  ]
}`